Signal/src/ViewControllers/MediaGalleryViewController.swift

```diff
@@ -0,0 +1,710 @@
+            self.presentationView.autoPinEdge(toSuperviewEdge: .left, withInset:convertedRect.origin.x)
+        ]
+    }
+
+    private func applyFinalMediaViewConstraints() {
+        if (self.presentationViewConstraints.count > 0) {
+            NSLayoutConstraint.deactivate(self.presentationViewConstraints)
+            self.presentationViewConstraints = []
+        }
+
+        self.presentationViewConstraints = [
+            self.presentationView.autoPinEdge(toSuperviewEdge: .leading),
+            self.presentationView.autoPinEdge(toSuperviewEdge: .top),
+            self.presentationView.autoPinEdge(toSuperviewEdge: .trailing),
+            self.presentationView.autoPinEdge(toSuperviewEdge: .bottom)
+        ]
+    }
+
+    private func applyOffscreenMediaViewConstraints() {
+        if (self.presentationViewConstraints.count > 0) {
+            NSLayoutConstraint.deactivate(self.presentationViewConstraints)
+            self.presentationViewConstraints = []
+        }
+
+        self.presentationViewConstraints += [
+            self.presentationView.autoPinEdge(toSuperviewEdge: .leading),
+            self.presentationView.autoPinEdge(toSuperviewEdge: .trailing),
+            self.presentationView.autoPinEdge(.top, to: .bottom, of: self.view)
+        ]
+    }
+
+    // MARK: MediaGalleryDataSource
+
+    lazy var mediaTileViewController: MediaTileViewController = {
+        let vc = MediaTileViewController(mediaGalleryDataSource: self, uiDatabaseConnection: self.uiDatabaseConnection)
+        vc.delegate = self
+        return vc
+    }()
+
+    var galleryItems: [MediaGalleryItem] = []
+    var sections: [GalleryDate: [MediaGalleryItem]] = [:]
+    var sectionDates: [GalleryDate] = []
+    var hasFetchedOldest = false
+    var hasFetchedMostRecent = false
+
+    func buildGalleryItem(message: TSMessage, transaction: YapDatabaseReadTransaction) -> MediaGalleryItem? {
+        guard let attachmentStream = message.attachment(with: transaction) as? TSAttachmentStream else {
+            owsFail("\(self.logTag) in \(#function) attachment was unexpectedly empty")
+            return nil
+        }
+
+        return MediaGalleryItem(message: message, attachmentStream: attachmentStream)
+    }
+
+    // Range instead of indexSet since it's contiguous?
+    var fetchedIndexSet = IndexSet() {
+        didSet {
+            Logger.debug("\(logTag) in \(#function) \(oldValue) -> \(fetchedIndexSet)")
+        }
+    }
+
+    func ensureGalleryItemsLoaded(_ direction: GalleryDirection, item: MediaGalleryItem, amount: UInt, completion: ((IndexSet, [IndexPath]) -> Void)? = nil ) {
+
+        var galleryItems: [MediaGalleryItem] = self.galleryItems
+        var sections: [GalleryDate: [MediaGalleryItem]] = self.sections
+        var sectionDates: [GalleryDate] = self.sectionDates
+
+        var newGalleryItems: [MediaGalleryItem] = []
+        var newDates: [GalleryDate] = []
+
+        Bench(title: "fetching gallery items") {
```

Comment:
I was going to remove these `Bench` calls before opening the PR, but they're kind of useful to see, and not in a hot code-path.

Reply:
Interesting.